src/ring_theory/dedekind_domain.lean

```diff
@@ -950,4 +950,586 @@ begin
+  exact hp (or.resolve_right ((irreducible_iff.1 hcontra).right p x hx') hx),
+end
+
+lemma pow_prime₂' {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M} (n : ℕ) (c : ℕ → M)
+  (h₁ : ∀ i j, i < j → dvd_not_unit (c i) (c j))
+  (h₂ : ∀ {r : M}, r ∣ q ↔ ∃ i ≤ n, associated r (c i)) : prime (c 1) :=
+begin
+  rw prime,
+  split,
+  { sorry },
+  { split,
+    { exact not_unit_of_dvd_not_unit (h₁ 0 1 zero_lt_one) },
+    { sorry } }, --the last part of the proof seems quite painful in the
+                -- `comm_cancel_monoid_with_zero M` setting
+end
+
+lemma pow_prime₂ {q : ideal T} (n : ℕ) (hn : 1 ≤ n) (c : ℕ → ideal T)
+  (h₁ : strict_anti c) (h₂ : ∀ (r : ideal T), r ∣ q ↔ ∃ i ≤ n, r = c i) :
+  (c 1).is_prime :=
+begin
+  apply is_maximal.is_prime,
+  rw [is_maximal_def, is_coatom],
+  split,
+  { exact ne_top_of_lt (h₁ zero_lt_one) },
+
+  intros b hb,
+  have h : b ∣ q := (dvd_iff_le.2 hb.le).trans ((h₂ (c 1)).2 ⟨1, hn, rfl⟩),
+  obtain ⟨i, hi, rfl⟩ := (h₂ b).1 h,
+  have H : i < 1 := h₁.lt_iff_lt.mp hb,
+  rw nat.lt_one_iff.mp H,
+  exact pow_prime₁ n c h₁ h₂
+end
+
+lemma is_unit_of_associated_is_unit {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M}
+  (h : associated p q) (hp : is_unit p): is_unit q :=
+by { obtain ⟨a, rfl⟩:= h, exact is_unit.mul hp (units.is_unit a) }
+
+lemma not_associated_of_dvd_not_unit {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M}
+  (h : dvd_not_unit p q) : ¬ associated p q :=
+begin
+  by_contra hcontra,
+  rw dvd_not_unit at h,
+  obtain ⟨hp, x, hx, hx'⟩ := h,
+  apply hx,
+  obtain ⟨a, rfl⟩ := hcontra,
+  rw mul_eq_mul_left_iff at hx',
+  rw ← or.resolve_right hx' hp,
+  exact units.is_unit a,
+end
+
+lemma dvd_not_unit_of_dvd_not_unit_associated {M : Type*} [comm_cancel_monoid_with_zero M]
+  {p q r : M} (h : dvd_not_unit p q) (h' : associated q r) : dvd_not_unit p r :=
+  sorry
+
+lemma pow_prime₃' {M : Type*} [comm_cancel_monoid_with_zero M] {p q r : M} (n : ℕ) (hn : 1 ≤ n)
+  (c : ℕ → M)  (h₁ : ∀ i j, i < j → dvd_not_unit (c i) (c j))
+  (h₂ : ∀ (r : M), r ∣ q ↔ ∃ i ≤ n,  associated r (c i)) (hp : prime p) (hr : r ∣ q)
+  (hr' : ¬ is_unit r) (hp' : p ∣ r) : associated p (c 1) :=
+begin
+  obtain ⟨i, hi, p_eq⟩ := (h₂ p).1 (dvd_trans hp' hr),
+  have : 1 ≤ i,
+  { rw [nat.succ_le_iff, pos_iff_ne_zero],
+    rintro rfl,
+    apply prime.not_unit hp,
+    apply is_unit_of_associated_is_unit (associated.symm p_eq) (pow_prime₁' n c h₁ h₂),
+    sorry, --Lean gives me a weird goal here, should go once `is_unit_of_associated_is_unit` is put in
+          -- the right file ?
+  },
+  by_cases h : 1 = i,             --this part is very messy
+  { rw ← h at p_eq,
+    exact p_eq },
+  { exfalso,
+    have : ¬ prime p,
+    { by_contra hcontra,
+      replace hcontra := prime.irreducible hcontra,
+      -- this part is quite messy. A clean-ish way of doing it would be to show that
+      -- `i ≥ 1 → ¬ unit (c i)` using induction
+      sorry,
+    },
+    exact this hp },
+end
```

Comment:
```suggestion

lemma pow_prime₃' {M : Type*} [comm_cancel_monoid_with_zero M] {p q r : associates M} (n : ℕ) (hn : 1 ≤ n)
  (c : ℕ → associates M)  (h₁ : strict_mono c)
  (h₂ : ∀ {r : associates M}, r ≤ q ↔ ∃ i ≤ n, r = c i) (hp : prime p) (hr : r ∣ q)
  (hr' : ¬ is_unit r) (hp' : p ∣ r) : p = c 1 :=
begin
  obtain ⟨i, hi, p_eq⟩ := h₂.1 (dvd_trans hp' hr),
  have : 1 ≤ i,
  { rw [nat.succ_le_iff, pos_iff_ne_zero],
    rintro rfl,
    apply prime.not_unit hp,
    rw p_eq,
    exact pow_prime₁' n c h₁ @h₂ },
  by_cases h : 1 = i,             --this part is very messy
  { rw ← h at p_eq,
    exact p_eq },
  { exfalso,
    have : ¬ prime p,
    { by_contra hcontra,
      replace hcontra := prime.irreducible hcontra,
      -- this part is quite messy. A clean-ish way of doing it would be to show that
      -- `i ≥ 1 → ¬ unit (c i)` using induction
      sorry,
    },
    exact this hp },
end
```